File: app/src/main/java/org/ole/planet/myplanet/model/RealmUserModel.java
package org.ole.planet.myplanet.model;

import android.content.SharedPreferences;
import android.text.TextUtils;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import org.apache.commons.lang3.StringUtils;
import org.ole.planet.myplanet.MainApplication;
import org.ole.planet.myplanet.utilities.JsonUtils;
import org.ole.planet.myplanet.utilities.NetworkUtils;
import org.ole.planet.myplanet.utilities.Utilities;
import org.ole.planet.myplanet.utilities.VersionUtils;

import java.util.Map;
import java.util.Set;
import java.util.UUID;

import io.realm.Realm;
import io.realm.RealmList;
import io.realm.RealmObject;
import io.realm.annotations.PrimaryKey;

public class RealmUserModel extends RealmObject {
    @PrimaryKey
    private String id;
    private String _id;
    private String _rev;
    private String name;
    private RealmList<String> roles;
    private Boolean isUserAdmin;
    private long joinDate;
    private String firstName;
    private String lastName;
    private String middleName;
    private String email;
    private String planetCode;
    private String parentCode;
    private String phoneNumber;
    private String password_scheme;
    private String iterations;
    private String derived_key;
    private String level;
    private String language;
    private String gender;
    private String salt;
    private String dob;
    private String birthPlace;
    private String communityName;
    private String userImage;
    private String key;
    private String iv;
    private String password;
    private boolean updated;
    private boolean showTopbar;

    public JsonObject serialize() {
        JsonObject object = new JsonObject();
        if (!get_id().isEmpty()) {
            object.addProperty("_id", get_id());
            object.addProperty("_rev", get_rev());
        }
        object.addProperty("name", getName());
        object.add("roles", getRoles());
        if (get_id().isEmpty()) {
            object.addProperty("password", getPassword());
            object.addProperty("macAddress", NetworkUtils.getMacAddr());
            object.addProperty("androidId",NetworkUtils.getMacAddr());
            object.addProperty("uniqueAndroidId",VersionUtils.getAndroidId(MainApplication.context));
            object.addProperty("customDeviceName", NetworkUtils.getCustomDeviceName(MainApplication.context));
        } else {
            object.addProperty("derived_key", getDerived_key());
            object.addProperty("salt", getSalt());
            object.addProperty("password_scheme", getPassword_scheme());
        }
        object.addProperty("isUserAdmin", getUserAdmin());
        object.addProperty("joinDate", getJoinDate());
        object.addProperty("firstName", getFirstName());
        object.addProperty("lastName", getLastName());
        object.addProperty("middleName", getMiddleName());
        object.addProperty("email", getEmail());
        object.addProperty("language", getLanguage());
        object.addProperty("level", getLevel());
        object.addProperty("type", "user");
        object.addProperty("gender", getGender());
        object.addProperty("phoneNumber", getPhoneNumber());
        object.addProperty("birthDate", getDob());
        try {
            object.addProperty("iterations", Integer.parseInt(getIterations()));
        }catch (Exception e){
            object.addProperty("iterations", 10);
        }
        object.addProperty("parentCode", getParentCode());
        object.addProperty("planetCode", getPlanetCode());
        object.addProperty("birthPlace", getBirthPlace());
        return object;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public String getKey() {
        return key;
    }

    public void setKey(String key) {
        this.key = key;
    }

    public String getIv() {
        return iv;
    }

    public boolean isUpdated() {
        return updated;
    }

    public void setUpdated(boolean updated) {
        this.updated = updated;
    }

    public void setIv(String iv) {
        this.iv = iv;
    }

    public String getLevel() {
        return level;
    }

    public void setLevel(String level) {
        this.level = level;
    }

    public String getLanguage() {
        return language;
    }

    public void setLanguage(String language) {
        this.language = language;
    }

    public String getGender() {
        return gender;
    }

    public void setGender(String gender) {
        this.gender = gender;
    }

    public static RealmUserModel createGuestUser(String username, Realm mRealm, SharedPreferences settings) {
        JsonObject object = new JsonObject();
        object.addProperty("_id", "guest_" + username);
        object.addProperty("name", username);
        object.addProperty("firstName", username);
        JsonArray rolesArray = new JsonArray();
        rolesArray.add("guest");
        object.add("roles", rolesArray);
        if (!mRealm.isInTransaction())
            mRealm.beginTransaction();
        return RealmUserModel.populateUsersTable(object, mRealm, settings);
    }

    public static RealmUserModel populateUsersTable(JsonObject jsonDoc, Realm mRealm, SharedPreferences settings) {
        try {
            String _id = JsonUtils.getString("_id", jsonDoc);
            if (_id.isEmpty())
                _id = UUID.randomUUID().toString();
            RealmUserModel user = mRealm.where(RealmUserModel.class).equalTo("_id", _id).findFirst();

            if (user == null) {
                user = mRealm.createObject(RealmUserModel.class, _id);
            }
            insertIntoUsers(jsonDoc, user, settings);

            return user;
        } catch (Exception err) {
            err.printStackTrace();
        }
        return null;
    }



    public static boolean isUserExists(Realm realm, String name) {
        return realm.where(RealmUserModel.class).equalTo("name", name).count() > 0;
    }

    private static void insertIntoUsers(JsonObject jsonDoc, RealmUserModel user, SharedPreferences settings) {
        Utilities.log("Insert into users " + new Gson().toJson(jsonDoc));
        user.set_rev(JsonUtils.getString("_rev", jsonDoc));
        user.set_id(JsonUtils.getString("_id", jsonDoc));
        user.setName(JsonUtils.getString("name", jsonDoc));
        JsonArray array = JsonUtils.getJsonArray("roles", jsonDoc);
        RealmList<String> roles = new RealmList<>();
        for (int i = 0; i < array.size(); i++) {
            roles.add(JsonUtils.getString(array, i));
        }
        user.setRoles(roles);
        user.setUserAdmin(JsonUtils.getBoolean("isUserAdmin", jsonDoc));
        user.setJoinDate(JsonUtils.getLong("joinDate", jsonDoc));
        user.setFirstName(JsonUtils.getString("firstName", jsonDoc));
        user.setLastName(JsonUtils.getString("lastName", jsonDoc));
        user.setMiddleName(JsonUtils.getString("middleName", jsonDoc));
        user.setPlanetCode(JsonUtils.getString("planetCode", jsonDoc));
        user.setParentCode(JsonUtils.getString("parentCode", jsonDoc));
        user.setEmail(JsonUtils.getString("email", jsonDoc));
        if (user.get_id().isEmpty())
            user.setPassword(JsonUtils.getString("password", jsonDoc));
        user.setPhoneNumber(JsonUtils.getString("phoneNumber", jsonDoc));
        user.setPassword_scheme(JsonUtils.getString("password_scheme", jsonDoc));
        user.setIterations(JsonUtils.getString("iterations", jsonDoc));
        user.setDerived_key(JsonUtils.getString("derived_key", jsonDoc));
        user.setSalt(JsonUtils.getString("salt", jsonDoc));
        user.setDob(JsonUtils.getString("birthDate", jsonDoc));
        user.setBirthPlace(JsonUtils.getString("birthPlace", jsonDoc));
        user.setGender(JsonUtils.getString("gender", jsonDoc));
        user.setLanguage(JsonUtils.getString("language", jsonDoc));
        user.setLevel(JsonUtils.getString("level", jsonDoc));
        user.setShowTopbar(true);
        user.addImageUrl(jsonDoc, settings);
        if (!TextUtils.isEmpty(JsonUtils.getString("planetCode", jsonDoc))) {
            settings.edit().putString("planetCode", JsonUtils.getString("planetCode", jsonDoc)).commit();
        }
        if (!TextUtils.isEmpty(JsonUtils.getString("parentCode", jsonDoc)))
            settings.edit().putString("parentCode", JsonUtils.getString("parentCode", jsonDoc)).commit();
    }

    public String getBirthPlace() {
        return birthPlace;
    }

    public void setBirthPlace(String birthPlace) {
        this.birthPlace = birthPlace;
    }

    public RealmList<String> getRolesList() {
        return roles;
    }

    public JsonArray getRoles() {
        JsonArray ar = new JsonArray();
        for (String s : roles
        ) {
            ar.add(s);
        }
        return ar;
    }

    public String get_id() {
        return _id;
    }

    public void set_id(String _id) {
        this._id = _id;
    }

    public void setRoles(RealmList<String> roles) {
        this.roles = roles;
    }

    public String getRoleAsString() {
        String s = StringUtils.join(getRolesList(), ",");
        return s;
    }

    public boolean isShowTopbar() {
        return showTopbar;
    }

    public void setShowTopbar(boolean showTopbar) {
        this.showTopbar = showTopbar;
    }

    public String getPlanetCode() {
        return planetCode;
    }

    public void setPlanetCode(String planetCode) {
        this.planetCode = planetCode;
    }

    public String getParentCode() {
        return parentCode;
    }

    public void setParentCode(String parentCode) {
        this.parentCode = parentCode;
    }

    public String getUserImage() {
        return userImage;
    }

    public void setUserImage(String userImage) {
        this.userImage = userImage;
    }

    public String getDob() {
        return dob;
    }

    public void setDob(String dob) {
        this.dob = dob;
    }

    public String getCommunityName() {
        return communityName;
    }

    public void setCommunityName(String communityName) {
        this.communityName = communityName;
    }

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String get_rev() {
        return _rev;
    }

    public void set_rev(String _rev) {
        this._rev = _rev;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getFullName() {
        return firstName + " " + lastName;
    }


    public Boolean getUserAdmin() {
        return isUserAdmin;
    }

    public void setUserAdmin(Boolean userAdmin) {
        isUserAdmin = userAdmin;
    }

    public long getJoinDate() {
        return joinDate;
    }

    public void setJoinDate(long joinDate) {
        this.joinDate = joinDate;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public String getMiddleName() {
        return middleName;
    }

    public void setMiddleName(String middleName) {
        this.middleName = middleName;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getPhoneNumber() {
        return phoneNumber;
    }

    public void setPhoneNumber(String phoneNumber) {
        this.phoneNumber = phoneNumber;
    }

    public String getPassword_scheme() {
        return password_scheme;
    }

    public void setPassword_scheme(String password_scheme) {
        this.password_scheme = password_scheme;
    }

    public String getIterations() {
        return iterations;
    }

    public void setIterations(String iterations) {
        this.iterations = iterations;
    }

    public String getDerived_key() {
        return derived_key;
    }

    public void setDerived_key(String derived_key) {
        this.derived_key = derived_key;
    }

    public String getSalt() {
        return salt;
    }

    public void setSalt(String salt) {
        this.salt = salt;
    }

    public void addImageUrl(JsonObject jsonDoc, SharedPreferences settings) {
        if (jsonDoc.has("_attachments")) {
            JsonParser parser = new JsonParser();
            JsonElement element = parser.parse(String.valueOf(jsonDoc.get("_attachments").getAsJsonObject()));
            JsonObject obj = element.getAsJsonObject();
            Set<Map.Entry<String, JsonElement>> entries = obj.entrySet();
            for (Map.Entry<String, JsonElement> entry : entries) {
                this.setUserImage(Utilities.getUserImageUrl(this.getId(), entry.getKey(), settings));
                break;
            }
        }
    }

    public Boolean getShowTopbar() {
        return showTopbar;
    }

    public void setShowTopbar(Boolean showTopbar) {
        this.showTopbar = showTopbar;
    }

    public boolean isManager() {
        JsonArray roles = getRoles();
        boolean isManager = roles.toString().toLowerCase().contains("manager") || isUserAdmin;
        return (isManager);
    }

    public boolean isLeader() {
        JsonArray roles = getRoles();
        return roles.toString().toLowerCase().contains("leader");
    }


    @Override
    public String toString() {
        return firstName + " " + lastName;
    }

}

File: app/src/main/java/org/ole/planet/myplanet/model/RealmUserModel.kt
package org.ole.planet.myplanet.model

import android.content.SharedPreferences
import android.util.Base64
import androidx.core.content.edit
import androidx.core.net.toUri
import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import io.realm.Realm
import io.realm.RealmList
import io.realm.RealmObject
import io.realm.annotations.PrimaryKey
import java.io.File
import java.io.InputStream
import java.util.Locale
import java.util.UUID
import org.apache.commons.lang3.StringUtils
import org.json.JSONException
import org.json.JSONObject
import org.ole.planet.myplanet.MainApplication.Companion.context
import org.ole.planet.myplanet.utilities.JsonUtils
import org.ole.planet.myplanet.utilities.NetworkUtils
import org.ole.planet.myplanet.utilities.UrlUtils
import org.ole.planet.myplanet.utilities.Utilities
import org.ole.planet.myplanet.utilities.VersionUtils

open class RealmUserModel : RealmObject() {
    @PrimaryKey
    var id: String? = null
    var _id: String? = null
    var _rev: String? = null
    var name: String? = null
    var rolesList: RealmList<String?>? = null
    var userAdmin: Boolean? = null
    var joinDate: Long = 0
    var firstName: String? = null
    var lastName: String? = null
    var middleName: String? = null
    var email: String? = null
    var planetCode: String? = null
    var parentCode: String? = null
    var phoneNumber: String? = null
    var password_scheme: String? = null
    var iterations: String? = null
    var derived_key: String? = null
    var level: String? = null
    var language: String? = null
    var gender: String? = null
    var salt: String? = null
    var dob: String? = null
    var age: String? = null
    var birthPlace: String? = null
    var userImage: String? = null
    var key: String? = null
    var iv: String? = null
    var password: String? = null
    var isUpdated = false
    var isShowTopbar = false
    var isArchived = false

    fun serialize(): JsonObject {
        val jsonObject = JsonObject()
        if (_id?.isNotEmpty() == true) {
            jsonObject.addProperty("_id", _id)
            jsonObject.addProperty("_rev", _rev)
        }
        jsonObject.addProperty("name", name)
        jsonObject.add("roles", getRoles())
        if (_id?.isEmpty() == true) {
            jsonObject.addProperty("password", password)
            jsonObject.addProperty("androidId", NetworkUtils.getUniqueIdentifier())
            jsonObject.addProperty("uniqueAndroidId", VersionUtils.getAndroidId(context))
            jsonObject.addProperty("customDeviceName", NetworkUtils.getCustomDeviceName(context))
        } else {
            jsonObject.addProperty("derived_key", derived_key)
            jsonObject.addProperty("salt", salt)
            jsonObject.addProperty("password_scheme", password_scheme)
        }
        jsonObject.addProperty("isUserAdmin", userAdmin)
        jsonObject.addProperty("joinDate", joinDate)
        jsonObject.addProperty("firstName", firstName)
        jsonObject.addProperty("lastName", lastName)
        jsonObject.addProperty("middleName", middleName)
        jsonObject.addProperty("email", email)
        jsonObject.addProperty("language", language)
        jsonObject.addProperty("level", level)
        jsonObject.addProperty("type", "user")
        jsonObject.addProperty("gender", gender)
        jsonObject.addProperty("phoneNumber", phoneNumber)
        jsonObject.addProperty("birthDate", dob)
        jsonObject.addProperty("age", age)
        try {
            jsonObject.addProperty("iterations", iterations?.takeIf { it.isNotBlank() }?.toInt() ?: 10)
        } catch (e: NumberFormatException) {
            e.printStackTrace()
            jsonObject.addProperty("iterations", 10)
        }
        jsonObject.addProperty("parentCode", parentCode)
        jsonObject.addProperty("planetCode", planetCode)
        jsonObject.addProperty("birthPlace", birthPlace)
        jsonObject.addProperty("isArchived", isArchived)

        val base64Image = encodeImageToBase64(userImage)

        if (!base64Image.isNullOrEmpty()) {
            val attachmentObject = JsonObject()
            val imageData = JsonObject()
            imageData.addProperty("content_type", "image/jpeg")
            imageData.addProperty("data", base64Image)

            attachmentObject.add("img", imageData)
            jsonObject.add("_attachments", attachmentObject)
        }

        return jsonObject
    }

    fun encodeImageToBase64(imagePath: String?): String? {
        if (imagePath.isNullOrEmpty()) return null
        return try {
            val inputStream: InputStream? = if (imagePath.startsWith("content://")) {
                val uri = imagePath.toUri()
                context.contentResolver.openInputStream(uri)
            } else {
                File(imagePath).inputStream()
            }

            inputStream?.use {
                val bytes = it.readBytes()
                Base64.encodeToString(bytes, Base64.NO_WRAP)
            }
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }

    private fun getRoles(): JsonArray {
        val ar = JsonArray()
        for (s in rolesList ?: emptyList())    {
            ar.add(s)
        }
        return ar
    }

    fun setRoles(roles: RealmList<String?>?) {
        rolesList = roles
    }

    fun getRoleAsString(): String {
        return StringUtils.join(rolesList, ",")
    }

    fun getFullName(): String {
        return "$firstName $lastName"
    }

    fun getFullNameWithMiddleName(): String {
        return "$firstName ${middleName ?: ""} $lastName"
    }

    fun addImageUrl(jsonDoc: JsonObject?) {
        if (jsonDoc?.has("_attachments") == true) {
            val element = JsonParser.parseString(jsonDoc["_attachments"].asJsonObject.toString())
            val obj = element.asJsonObject
            val entries = obj.entrySet()
            for ((key1) in entries) {
                userImage = UrlUtils.getUserImageUrl(id, key1)
                break
            }
        }
    }

    fun isManager(): Boolean {
        val roles = getRoles()
        val isManager = roles.toString().lowercase(Locale.ROOT).contains("manager") || userAdmin ?: false
        return isManager
    }

    fun isLeader(): Boolean {
        val roles = getRoles()
        return roles.toString().lowercase(Locale.ROOT).contains("leader")
    }

    fun isGuest(): Boolean {
        val hasGuestId = _id?.startsWith("guest_") == true
        val hasGuestRole = rolesList?.any { it?.lowercase() == "guest" } == true
        return hasGuestId || (hasGuestRole && rolesList?.any { it?.lowercase() == "learner" } != true)
    }

    override fun toString(): String {
        return "$name"
    }

    companion object {
        @JvmStatic
        fun createGuestUser(username: String?, mRealm: Realm, settings: SharedPreferences): RealmUserModel? {
            val `object` = JsonObject()
            `object`.addProperty("_id", "guest_$username")
            `object`.addProperty("name", username)
            `object`.addProperty("firstName", username)
            val rolesArray = JsonArray()
            rolesArray.add("guest")
            `object`.add("roles", rolesArray)
            var managedGuest: RealmUserModel? = null
            val shouldExecuteTransaction = !mRealm.isInTransaction

            return try {
                if (shouldExecuteTransaction) {
                    mRealm.executeTransaction { realm ->
                        managedGuest = populateUsersTable(`object`, realm, settings)
                    }
                } else {
                    managedGuest = populateUsersTable(`object`, mRealm, settings)
                }
                managedGuest?.let { mRealm.copyFromRealm(it) }
            } catch (err: Exception) {
                if (shouldExecuteTransaction && mRealm.isInTransaction) {
                    mRealm.cancelTransaction()
                }
                err.printStackTrace()
                null
            }
        }

        @JvmStatic
        fun populateUsersTable(jsonDoc: JsonObject?, mRealm: Realm?, settings: SharedPreferences): RealmUserModel? {
            if (jsonDoc == null || mRealm == null) return null
            try {
                val id = JsonUtils.getString("_id", jsonDoc).takeIf { it.isNotEmpty() } ?: UUID.randomUUID().toString()
                val userName = JsonUtils.getString("name", jsonDoc)
                var user: RealmUserModel? = null

                if (!mRealm.isInTransaction) {
                    mRealm.executeTransaction { realm ->
                        user = realm.where(RealmUserModel::class.java)
                            .equalTo("_id", id)
                            .findFirst()

                        if (user == null && id.startsWith("org.couchdb.user:") && userName.isNotEmpty()) {
                            val guestUser = realm.where(RealmUserModel::class.java)
                                .equalTo("name", userName)
                                .beginsWith("_id", "guest_")
                                .findFirst()

                            if (guestUser != null) {
                                val tempData = JsonObject()
                                tempData.addProperty("_id", id)
                                tempData.addProperty("name", guestUser.name)
                                tempData.addProperty("firstName", guestUser.firstName)
                                tempData.addProperty("lastName", guestUser.lastName)
                                tempData.addProperty("middleName", guestUser.middleName)
                                tempData.addProperty("email", guestUser.email)
                                tempData.addProperty("phoneNumber", guestUser.phoneNumber)
                                tempData.addProperty("level", guestUser.level)
                                tempData.addProperty("language", guestUser.language)
                                tempData.addProperty("gender", guestUser.gender)
                                tempData.addProperty("birthDate", guestUser.dob)
                                tempData.addProperty("planetCode", guestUser.planetCode)
                                tempData.addProperty("parentCode", guestUser.parentCode)
                                tempData.addProperty("userImage", guestUser.userImage)
                                tempData.addProperty("joinDate", guestUser.joinDate)
                                tempData.addProperty("isShowTopbar", guestUser.isShowTopbar)
                                tempData.addProperty("isArchived", guestUser.isArchived)

                                val rolesArray = JsonArray()
                                guestUser.rolesList?.forEach { role ->
                                    rolesArray.add(role)
                                }
                                tempData.add("roles", rolesArray)
                                guestUser.deleteFromRealm()
                                user = realm.createObject(RealmUserModel::class.java, id)
                                user?.let { insertIntoUsers(tempData, it, settings) }
                            }
                        }

                        if (user == null) {
                            user = realm.createObject(RealmUserModel::class.java, id)
                        }
                        user?.let { insertIntoUsers(jsonDoc, it, settings) }
                    }
                } else {
                    user = mRealm.where(RealmUserModel::class.java)
                        .equalTo("_id", id)
                        .findFirst()

                    if (user == null && id.startsWith("org.couchdb.user:") && userName.isNotEmpty()) {
                        val guestUser = mRealm.where(RealmUserModel::class.java)
                            .equalTo("name", userName)
                            .beginsWith("_id", "guest_")
                            .findFirst()

                        if (guestUser != null) {
                            val tempData = JsonObject()
                            tempData.addProperty("_id", id)
                            tempData.addProperty("name", guestUser.name)
                            tempData.addProperty("firstName", guestUser.firstName)
                            tempData.addProperty("lastName", guestUser.lastName)
                            tempData.addProperty("middleName", guestUser.middleName)
                            tempData.addProperty("email", guestUser.email)
                            tempData.addProperty("phoneNumber", guestUser.phoneNumber)
                            tempData.addProperty("level", guestUser.level)
                            tempData.addProperty("language", guestUser.language)
                            tempData.addProperty("gender", guestUser.gender)
                            tempData.addProperty("birthDate", guestUser.dob)
                            tempData.addProperty("planetCode", guestUser.planetCode)
                            tempData.addProperty("parentCode", guestUser.parentCode)
                            tempData.addProperty("userImage", guestUser.userImage)
                            tempData.addProperty("joinDate", guestUser.joinDate)
                            tempData.addProperty("isShowTopbar", guestUser.isShowTopbar)
                            tempData.addProperty("isArchived", guestUser.isArchived)
                            val rolesArray = JsonArray()
                            guestUser.rolesList?.forEach { role ->
                                rolesArray.add(role)
                            }
                            tempData.add("roles", rolesArray)
                            guestUser.deleteFromRealm()
                            user = mRealm.createObject(RealmUserModel::class.java, id)
                            user?.let { insertIntoUsers(tempData, it, settings) }
                        }
                    }

                    if (user == null) {
                        user = mRealm.createObject(RealmUserModel::class.java, id)
                    }
                    user?.let { insertIntoUsers(jsonDoc, it, settings) }
                }
                return user
            } catch (err: Exception) {
                err.printStackTrace()
            }
            return null
        }

        private fun insertIntoUsers(jsonDoc: JsonObject?, user: RealmUserModel, settings: SharedPreferences) {
            if (jsonDoc == null) return

            val planetCodes = JsonUtils.getString("planetCode", jsonDoc)
            val rolesArray = JsonUtils.getJsonArray("roles", jsonDoc)
            val newId = JsonUtils.getString("_id", jsonDoc)

            user.apply {
                _rev = JsonUtils.getString("_rev", jsonDoc)
                _id = newId
                name = JsonUtils.getString("name", jsonDoc)
                setRoles(RealmList<String?>().apply {
                    for (i in 0 until rolesArray.size()) {
                        add(JsonUtils.getString(rolesArray, i))
                    }
                })
                userAdmin = JsonUtils.getBoolean("isUserAdmin", jsonDoc)
                val newJoinDate = JsonUtils.getLong("joinDate", jsonDoc)
                if (newJoinDate != 0L || joinDate == 0L) {
                    joinDate = newJoinDate
                }

                val newFirstName = JsonUtils.getString("firstName", jsonDoc)
                if (newFirstName.isNotEmpty() || firstName.isNullOrEmpty()) {
                    firstName = newFirstName
                }

                val newLastName = JsonUtils.getString("lastName", jsonDoc)
                if (newLastName.isNotEmpty() || lastName.isNullOrEmpty()) {
                    lastName = newLastName
                }

                val newMiddleName = JsonUtils.getString("middleName", jsonDoc)
                if (newMiddleName.isNotEmpty() || middleName.isNullOrEmpty()) {
                    middleName = newMiddleName
                }

                val newEmail = JsonUtils.getString("email", jsonDoc)
                if (newEmail.isNotEmpty() || email.isNullOrEmpty()) {
                    email = newEmail
                }

                val newPhoneNumber = JsonUtils.getString("phoneNumber", jsonDoc)
                if (newPhoneNumber.isNotEmpty() || phoneNumber.isNullOrEmpty()) {
                    phoneNumber = newPhoneNumber
                }

                val newLevel = JsonUtils.getString("level", jsonDoc)
                if (newLevel.isNotEmpty() || level.isNullOrEmpty()) {
                    level = newLevel
                }

                val newLanguage = JsonUtils.getString("language", jsonDoc)
                if (newLanguage.isNotEmpty() || language.isNullOrEmpty()) {
                    language = newLanguage
                }

                val newGender = JsonUtils.getString("gender", jsonDoc)
                if (newGender.isNotEmpty() || gender.isNullOrEmpty()) {
                    gender = newGender
                }

                val newDob = JsonUtils.getString("birthDate", jsonDoc)
                if (newDob.isNotEmpty() || dob.isNullOrEmpty()) {
                    dob = newDob
                }

                val newBirthPlace = JsonUtils.getString("birthPlace", jsonDoc)
                if (newBirthPlace.isNotEmpty() || birthPlace.isNullOrEmpty()) {
                    birthPlace = newBirthPlace
                }

                val newAge = JsonUtils.getString("age", jsonDoc)
                if (newAge.isNotEmpty() || age.isNullOrEmpty()) {
                    age = newAge
                }
                planetCode = planetCodes
                parentCode = JsonUtils.getString("parentCode", jsonDoc)
                if (_id?.isEmpty() == true) {
                    password = JsonUtils.getString("password", jsonDoc)
                }
                password_scheme = JsonUtils.getString("password_scheme", jsonDoc)
                iterations = JsonUtils.getString("iterations", jsonDoc)
                derived_key = JsonUtils.getString("derived_key", jsonDoc)
                salt = JsonUtils.getString("salt", jsonDoc)
                isShowTopbar = true
                isArchived = JsonUtils.getBoolean("isArchived", jsonDoc)
                addImageUrl(jsonDoc)
            }

            if (planetCodes.isNotEmpty()) {
                settings.edit { putString("planetCode", planetCodes) }
            }
        }

        @JvmStatic
        fun isUserExists(realm: Realm, name: String?): Boolean {
            return realm.where(RealmUserModel::class.java)
                .equalTo("name", name)
                .not().beginsWith("_id", "guest").count() > 0
        }

        fun updateUserDetails(realm: Realm, userId: String?, firstName: String?, lastName: String?,
        middleName: String?, email: String?, phoneNumber: String?, level: String?, language: String?,
        gender: String?, dob: String?, onSuccess: () -> Unit) {
            realm.executeTransactionAsync({ mRealm ->
                val user = mRealm.where(RealmUserModel::class.java).equalTo("id", userId).findFirst()
                if (user != null) {
                    user.firstName = firstName
                    user.lastName = lastName
                    user.middleName = middleName
                    user.email = email
                    user.phoneNumber = phoneNumber
                    user.level = level
                    user.language = language
                    user.gender = gender
                    user.dob = dob
                    user.isUpdated = true
                }
            }, {
                onSuccess.invoke()
                Utilities.toast(context, "User details updated successfully")
            }) {
                Utilities.toast(context, "User details update failed")
            }
        }

        @JvmStatic
        fun parseLeadersJson(jsonString: String): List<RealmUserModel> {
            val leadersList = mutableListOf<RealmUserModel>()
            try {
                val jsonObject = JSONObject(jsonString)
                val docsArray = jsonObject.getJSONArray("docs")
                for (i in 0 until docsArray.length()) {
                    val docObject = docsArray.getJSONObject(i)
                    val user = RealmUserModel()
                    user.name = docObject.getString("name")
                    if (!docObject.isNull("firstName")) {
                        user.firstName = docObject.getString("firstName")
                    }
                    if (!docObject.isNull("lastName")) {
                        user.lastName = docObject.getString("lastName")
                    }
                    if (!docObject.isNull("email")) {
                        user.email = docObject.getString("email")
                    }
                    leadersList.add(user)
                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            return leadersList
        }

        @JvmStatic
        fun cleanupDuplicateUsers(realm: Realm, onSuccess: () -> Unit) {
            realm.executeTransactionAsync({ mRealm: Realm ->
                val allUsers = mRealm.where(RealmUserModel::class.java).findAll()
                val usersByName = allUsers.groupBy { it.name }

                usersByName.forEach { (_, users) ->
                    if (users.size > 1) {
                        val sortedUsers = users.sortedWith { user1, user2 ->
                            when {
                                user1._id?.startsWith("org.couchdb.user:") == true &&
                                        user2._id?.startsWith("guest_") == true -> -1
                                user1._id?.startsWith("guest_") == true &&
                                        user2._id?.startsWith("org.couchdb.user:") == true -> 1
                                else -> 0
                            }
                        }

                        for (i in 1 until sortedUsers.size) {
                            sortedUsers[i].deleteFromRealm()
                        }
                    }
                }
            }, {
                onSuccess.invoke()
            }) {
            }
        }
    }
}
